salt/client/ssh/__init__.py

```diff
@@ -582,11 +596,40 @@ def handle_ssh(self, mine=False):
                 log.error("No matching targets found in roster.")
                 break
             if len(running) < self.opts.get("ssh_max_procs", 25) and not init:
-                try:
-                    host = next(target_iter)
-                except StopIteration:
+                if targets_queue:
+                    host = targets_queue.pop(0)
+                else:
                     init = True
                     continue
+                with salt.utils.files.flopen(self.session_flock_file, "w"):
+                    cached_session = self.cache.fetch("salt-ssh/session", host)
+                    if cached_session is not None and "ts" in cached_session:
+                        prev_session_running = time.time() - cached_session["ts"]
+                        if (
+                            "pid" in cached_session
+                            and cached_session.get("master_id", self.master_id)
+                            == self.master_id
+                        ):
+                            pid_running = (
+                                False
+                                if cached_session["pid"] == 0
+                                else psutil.pid_exists(cached_session["pid"])
+                            )
+                            if (
+                                pid_running and prev_session_running < self.max_pid_wait
+                            ) or (not pid_running and prev_session_running < 3):
```

Comment:
Should we make this grace time of `3` configurable somehow?

Reply:
Sure, we can. Maybe not very useful, but it could help if the environment has some network issues. Without this waiting the chances to get error with unavailability to open port forwarding are much higher. It seems sshd on the minion side need some time to bring the port down

Reply:
Made configurable with https://github.com/openSUSE/salt/pull/476/commits/5ef5d28051030f2607f78d8f17d359bf66bd6dc0